### ISSUE TITLE
`RuntimeError: out of bounds memory access` when using plugin on a large nested AST

### ISSUE THREAD

**Issue by jzhan-canva** [swc-project/swc#9782]
### Describe the bug

When a large nested file being loaded to swc with plugin, the plugin_runner will crash.
When same file being loaded to swc without plugin, no error is thrown.

### Input code

```typescript
export result = {
      output: outputBuilder()
        .addConfig(1)
        .addConfig(2)
        .addConfig(3)
        .addConfig(4)
        .addConfig(5)
        ....
        .addConfig(1500)
        .build(),
    }
```

### Config

```json
{
  "$schema": "https://swc.rs/schema.json",
  "jsc": {
    "experimental": {
      "plugins": [["@swc/plugin-noop", {}]]
    }
  }
}
```

### Playground link (or link to the minimal reproduction)

https://github.com/jzhan-canva/swc-plugin-bug

### Expected behavior

Same a running swc without plugin, that it should run

### Actual behavior

Crash
![image](https://github.com/user-attachments/assets/5d45d3ec-9e3c-434b-a209-3b3818d940cf)

### Version

"@swc/cli": "^0.5.2",     "@swc/core": "^1.9.3",     "@swc/plugin-noop": "^5.0.0"

**Comment by jzhan-canva** [comment]
upon further investigation, it seems the bug only happen when AST tree is too deep, not the size
for example in the reproducible example, I have a builder with 1000+ depth member expression, and it trigger the issue
but if I reduce depth, I can duplicate the builder code many times without any issue

**Comment by kdy1** [comment]
So my guess is that it's about stack overflow rather than a memory usage. Stack sizes are limited, and I think `rkyv` is failing due to the nested call level being too deep.

**Comment by kwonoj** [comment]
I believe we have same issues in the host as well? if depth is deep enough, either host or plugin won't able to traverse down with visitor.

**Comment by jzhan-canva** [comment]
Right now the host and guest have different limit. I think it makes sense to have same limit for host and guest

**Comment by jzhan-canva** [comment]
Current host limit is bigger, so the identical file can run without plugin but not with plugin

**Comment by jzhan-canva** [comment]
Current host limit is bigger, so the identical file can run without plugin but not with plugin

**Comment by kwonoj** [comment]
> I think it makes sense to have same limit for host and guest

I'm not sure if this is configurable on our end, ideally hope to do so.

**Comment by kdy1** [comment]
No, I mean it's failing with `rkyv`. The example plugin does not visit the program.

**Comment by jzhan-canva** [comment]
I can find this similar issue #1627 and it was resolved by #8867 
maybe this `maybe_grow_default` is needed in `rkyv` as well?

**Comment by jzhan-canva** [comment]
I'm able to reproduce this issue in [a unit test](https://github.com/swc-project/swc/compare/main...jzhan-canva:jzhan-unit-test-9782?expand=1)

**Comment by bradzacher** [comment]
is there a way that we can workaround this?
for example -- could we pull in SWC into a rust crate to include the plugin natively, rather than as WASM?

**Comment by kdy1** [comment]
Yes, and I'm actively maintaining it, although it's not in the current repository and there's no clean API for it as it's not designed for a general use.
`next-swc` is `@swc/core` with some css-in-js plugins.

The base code of a custom SWC binary with `transform()` lives at https://github.com/swc-project/swc/blob/5a584078645b8d62df5591a1938be3f0d669c544/bindings/binding_core_node/src/transform.rs and `next-swc` defines a function named `custom_before_pass`, and creates a sequence of all custom passes. https://github.com/vercel/next.js/blob/5f455e1a9533f8a29076a30edbaf87a0cccdcdc0/crates/next-custom-transforms/src/chain_transforms.rs#L121-L329

After declaring your own chain of the custom passes, you can pass it to `process_js_with_custom_pass` like https://github.com/vercel/next.js/blob/5f455e1a9533f8a29076a30edbaf87a0cccdcdc0/crates/napi/src/transform.rs#L138-L147

**Comment by kdy1** [comment]
@bradzacher Please tell me if you need more guidance

**Comment by jzhan-canva** [comment]
To follow up on this issue, I did a quick test on latest swc 
```
    "@swc/cli": "^0.7.3",
    "@swc/core": "^1.11.21",
    "@swc/plugin-noop": "^7.0.3"
```
I can't reproduce the original error with same input.js file, it seems the newer version has improved the overflow issue
However, when I keep adding depth to the input.js file, it causes the plugin system to fail again (building without plugin is still fine)

<img width="902" alt="Image" src="https://github.com/user-attachments/assets/39922021-f583-49e2-a14f-2121619a8dd4" />

**Comment by kwonoj** [comment]
Afaik this may be related https://github.com/swc-project/swc/pull/10359. I think there's hard limit at some point, so keep adding stack eventually will cause oom at some point.

**Comment by jzhan-canva** [comment]
Thanks for the information, I'm going to test whether current stack size is enough for our use case

**Comment by kdy1** [comment]
@jzhan-canva Can you try changes in https://github.com/swc-project/swc/pull/10359/files#diff-d1319bb6ecbabc24b029b4016b78bb823c4f0577dba4e939256b182ef239f4aa ?
